### ISSUE TITLE
K8s files are not scannable (not valid YAML)

### ISSUE THREAD

**Issue by gurjmatharu** [bcgov/cas-cif#331]
The helm chart is not valid yaml.  After substitution it becomes valid and deployable, however this does not allow us to run any tooling on these charts.

This issue needs further discussion, we are currently unable to run any of the following:

- [kubeaudit](https://github.com/Shopify/kubeaudit)
- [checkov](https://www.checkov.io)
- [snyk](https://docs.snyk.io/products/snyk-infrastructure-as-code/scan-kubernetes-configuration-files/scan-and-fix-security-issues-in-helm-charts)

**Comment by matthieu-foucault** [comment]
The output of the `helm template` command, used in the [`lint_chart` target](https://github.com/bcgov/cas-cif/blob/e9dd6661963057b56842132447cc86080bd2b8c9/Makefile#L203), should be valid k8s object, so that should be a valid input for the tools you mentioned above

**Comment by gurjmatharu** [comment]
Great! We should still integrate these tools into our pipeline.

**Comment by gurjmatharu** [comment]
Additionally we may want to look into GitOps.  
Some resources:
- https://cloud.google.com/kubernetes-engine/docs/tutorials/gitops-cloud-build
- https://www.weave.works/technologies/gitops/

**Comment by matthieu-foucault** [comment]
I'm not opposed to adding more checks in our CI, but I'd like more information on:
- what issues those tools check for
- do they require and account creation, and if so are there any costs

As per #306 , I'd like to make sure that security checks failing are not blocking daily development activities. There might be some nuances to discuss there, e.g. we might want critical security issues to be blocking to ensure they get attention, and we'll have to define how we tackle other issues, to ensure that re improve our application security.

> Additionally we may want to look into GitOps.
You'll have to be more precise here. Other than some initial setup for our namespaces, Git is already the source of truth for our CI/CD pipeline, so I'm not sure what you mean

**Comment by gurjmatharu** [comment]
They're free open sourced tooling - we wouldn't use them all just one is sufficient.  For example kubeaudit:

> kubeaudit is a command line tool and a Go package to audit Kubernetes clusters for various different security concerns, such as:
> 
> run as non-root
> use a read-only root filesystem
> drop scary capabilities, don't add new ones
> don't run privileged
> and more!
> tldr. kubeaudit makes sure you deploy secure containers!

As for more checks this comes down to our risk appetite.  Our security posture is exceptional but stops at the base container image and doesn't seem to include the application deployment itself.

Gitops
Although it's definitely the source of truth our cluster is not declarative.  At any given time we cannot refer to our git repository to have an understanding of the state of our deployments on our cluster.  Unless I am mistaken of course. The idea here is to have our infrastructure as code in a declarative way.   
Please refer to https://www.weave.works/technologies/gitops/ for GitOps in the context of k8s land.
There isn't quite a quick fix for this one - worth chatting about though.

**Comment by matthieu-foucault** [comment]
>run as non-root
use a read-only root filesystem
drop scary capabilities, don't add new ones
don't run privileged

Those examples are not something we have control over in our shared OpenShift cluster, so that's something to consider when looking at those tools.

On the GitOps side we're also limited by what's available to us. Argo CD was recently added to the cluster, so that's something that can be explored.

**Comment by gurjmatharu** [comment]
Sounds good - we do have control over our manifest files which this tool also audits. See here: https://github.com/Shopify/kubeaudit/blob/main/docs/autofix.md

ArgoCD would be perfect. (if we decide to go down this route)

**Comment by pbastia** [comment]
Reach out to get a demo of ArgoCD in the early year

**Comment by pbastia** [comment]
Junmin also mentioned looking into Tekton for the CI needs directly on the cluster.

**Comment by pbastia** [comment]
There is some documentation here: https://github.com/BCDevOps/openshift-wiki/tree/master/docs/ArgoCD
